### PR TITLE
Feature/odp

### PR DESCRIPTION
**Pull request by RKJS689**
- Adding the ODP client library method to enable customers to use the incoming ODP features.
- Test file for the ODP client library 
- Addition of a dedicated change log file

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -0,0 +1,98 @@
+Changelog
+=========================
+##v.2.3.1
```
Comment (dugjason) [inline CHANGELOG.md]: This needs to be v.2.4.0 as we are adding a feature

**Comment by dugjason** [comment]
We need to get the User Agent string updated: https://github.com/datasift/datasift-php/blob/21f8cc7649bfda38f180b7d4fba8082dee01e176/lib/DataSift/User.php#L30

**Comment by paulmozo** [comment]
I'll update all the version stuff in the develop branch before release, everything else looks alright.

**Comment by dugjason** [comment]
Receiving parse error when running ODP_Example:

```
Parse error: parse error, expecting `')'' in ~/datasift-php/examples/ODP_Example.php on line 27
```

File: lib/DataSift/ApiClient.php [hunk 2]
```diff
@@ -117,6 +122,12 @@ static private function initialize($method, $ssl, $url, $headers, $params, $user
                 curl_setopt($ch, CURLOPT_POSTFIELDS, json_encode($params));
                 break;
             }
+
+            case 'ingest': {
```
Comment (datasift-chrisknight) [inline lib/DataSift/ApiClient.php]: These in the switch statement are to do with which type of REST call is being done, it should only contain, GET, POST, PUT, PATCH and DELETE. This ingest method is the same as POST

File: lib/DataSift/ApiClient.php [hunk 3]
```diff
@@ -141,6 +152,8 @@ static private function initialize($method, $ssl, $url, $headers, $params, $user
 
         $url = self::appendQueryString($url, $qs);
 
+        var_dump($url);
```
Comment (datasift-chrisknight) [inline lib/DataSift/ApiClient.php]: Debugging?

File: lib/DataSift/ApiClient.php [hunk 4]
```diff
@@ -67,7 +67,12 @@ static public function call(
         $ssl = $user->useSSL();
 
         // Build the full endpoint URL
-        $url = 'http'.($ssl ? 's' : '').'://'.$user->getApiUrl(). $user->getApiVersion() . '/'. $endPoint;
+        if ($method == 'ingest') {
```
Comment (datasift-chrisknight) [inline lib/DataSift/ApiClient.php]: I'm not overly keen on this, should this not work more like how the streaming_url model works? Should the ingestion of ODP be done with a specific endpoint related. I'm not sure hard coding exceptions like this is the right way to be doing it.

File: lib/DataSift/ODP.php [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+	*/
+	public function getSourceId()
+	{
+		return $this->_source_id;
+	}
+
+	/**
+	* Generates a curl request to the Ingestion Endpoint 
+	*/
+	public function ingest($data_set)
```
Comment (datasift-chrisknight) [inline lib/DataSift/ODP.php]: This isn't necessary, should be using post

File: lib/DataSift/User.php [hunk 6]
```diff
@@ -524,6 +544,35 @@ private function handleResponse($res)
...
+     * Make a call to a DataSift ingestion endpoint.
+     *
+     * @param string $endpoint The endpoint of the API call.
+     * @param array  $params   The parameters to be passed along with the request.
+     *
+     * @return array The response from the server.
+     * @throws DataSift_Exception_APIError
+     * @throws DataSift_Exception_RateLimitExceeded
+     */
+    public function ingest($endpoint, $params = array(), $headers = array())
```
Comment (datasift-chrisknight) [inline lib/DataSift/User.php]: This isn't necessary, should be using post.

File: lib/DataSift/ApiClient.php (hunk 4 above)
Comment (paulmozo) [inline lib/DataSift/ApiClient.php]: @datasift-chrisknight You're right it's not exactly nice but it was an easy solution to this given the current structure of the library. Happy to talk it through if you have an idea on how to handle multipel URLs

Comment (datasift-chrisknight) [inline lib/DataSift/ApiClient.php]: Surely we could pass in a flag on the call? If it's flagged as ingestion we swap the url on initialise? Then we don't need these extra additions to the methods, which I've flagged below?

Comment (paulmozo) [inline lib/DataSift/ApiClient.php]: Yeah that could work, I'll take a look tomorrow unless you wanted to take a look.

Comment (datasift-chrisknight) [inline lib/DataSift/ApiClient.php]: I wish, I have much Billing Pipeline to replace!! :)

**Comment by paulmozo** [comment]
@datasift-chrisknight Thoughts on those changes?

**Comment by datasift-chrisknight** [comment]
Looks good @paulmozo

**Pull request merged**